### PR TITLE
File HANDLE/descriptor translation layer for Windows

### PR DESCRIPTION
**Pull request by mehrdadn**
## Why are these changes needed?

Windows uses `HANDLE`s for many things, which are pointer-sized, but much of our code & external libraries uses FDs, which are `int`-sized.

While the C runtime (CRT) on Windows supports FDs as well, the various APIs are not drop-in substitutes. Sockets require special treatment, and, moreover, the CRT rejects some `HANDLE`s (like file mappings to memory) and does not allow them to be placed inside the FD table.

This PR plays some tricks to provide an almost-seamless translation layer between FDs and `HANDLE`s on Windows, using the existing FD table for the underlying functionality as much as possible.

## Related issue number

#631

## Checks

- [ ] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://ray.readthedocs.io/en/latest/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failure rates at https://ray-travis-tracker.herokuapp.com/.
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested (please justify below)

### CODE REVIEW

File: bazel/ray_deps_setup.bzl [hunk 1]
```diff
@@ -171,11 +167,16 @@ def ray_deps_setup():
...
             "//thirdparty/patches:arrow-headers-unused.patch",
             "//thirdparty/patches:arrow-windows-export.patch",
-            "//thirdparty/patches:arrow-windows-poll.patch",
+            "//thirdparty/patches:arrow-windows-nonstdc.patch",
             "//thirdparty/patches:arrow-windows-sigpipe.patch",
             "//thirdparty/patches:arrow-windows-socket.patch",
             "//thirdparty/patches:arrow-windows-dlmalloc.patch",
+            "//thirdparty/patches:arrow-windows-tcp.patch",
         ],
+    ) if True else native.new_local_repository(
```
Comment (pcmoritz) [inline bazel/ray_deps_setup.bzl]: we should remove this before merging :)

Comment (mehrdadn) [reply bazel/ray_deps_setup.bzl]: Oops, good catch! Just pushed a fix :)

File: python/ray/node.py [hunk 2]
```diff
@@ -397,14 +398,20 @@ def _prepare_socket_file(self, socket_path, default_prefix):
...
-        return self._make_inc_temp(
-            prefix=default_prefix, directory_name=self._sockets_dir)
+            if sys.platform != "win32":
+                if os.path.exists(socket_path):
+                    raise RuntimeError(
+                        "Socket file {} exists!".format(socket_path))
+                socket_dir = os.path.dirname(socket_path)
+                try_to_create_directory(socket_dir)
+            result = socket_path
+        elif sys.platform == "win32":
```
Comment (pcmoritz) [inline python/ray/node.py]: Somehow the nesting structure of the ifs is not super clear at the moment. I think better would be:
```
if not windows:
  if socket_path is None:
    # create_or_runtime_error and return
  else:
    return self._make_ind_temp(...)
else:
  return socket_path or "tcp://{}:{}".format(self._localhost, self._get_unused_port())
```

What do you think?

Comment (mehrdadn) [reply python/ray/node.py]: I agree that looks better, but it also changes the behavior when `socket_path` is e.g. an empty string. Is that fine?

Comment (mehrdadn) [reply python/ray/node.py]: Although maybe it'll still be more readable if I mirror the top part with `if socket_path is None`, I could try that.

**Review by pcmoritz — APPROVED** [review]

**Pull request merged**